core/ir/instr_api.h

```diff
@@ -2337,6 +2337,11 @@ DR_API
 instr_t *
 instr_create_pusha(void *drcontext);
 
+DR_API
+/** Returns whether \p instr loads mcontext base to some reg. */
+bool
+instr_is_load_mcontext_base(instr_t *instr);
+
```

Comment:
Re-starting a thread here where we have better threading and tracking:

Could we avoid drreg having to recognize *all* register moves, including DR's,  with just a couple of label instructions added to the ilist at slot-changing points or something?  Trying to figure out what the alternative are.

Reply:
I tried to include some relevant examples below.

At the very least, we need to be aware of app instrs. Consider this case where we have multiple restores (which happens when we restore the reg before app read instrs):

example1:
```
<spill reg r1 to s1>
<restore reg r1 from s1>
<restore reg r1 from s1>
<crash>
```

Now, after the crash, do we restore r1 from s1 or not? If s1 still has r1’s app value, we can restore it (and live with a redundant restore in some cases), but if r1 was written by some app instruction before the crash, then s1 is stale and we should not restore. So we definitely need to look at app instrs. Whenever we have an app write, we forget all spill slots for that reg.

The next question is: can we avoid looking at meta instrs and the need to recognize non-drreg spills. In the current drreg restore logic, this is needed to be able to differentiate between tool val spills and app val spills. Without knowing whether a reg contains its native value currently, it’s hard to do that. We have used other heuristics in the past: like checking whether the reg was already spilled to some slot, and if so, assuming that the next spill is a tool val spill, but that has its limitations -- e.g. in some cases we may need to remember the second spill slot, like example2 below.

example2:
```
<app2app spills app r1 to s1>
<insertion phase spills app r1 to s2>
<app2app restores app r1 from s1>
<insertion phase writes some tool val to r1>
<app2app spills app r2 to s1>
// at this point app r1 is only present in s2
```



For aflags, this becomes a tougher problem. Because we don’t even keep the same slot at re-spills for aflags, or a slot at all (if spilled to xax). So we cannot even rely on each phase using the same slot always.

Coming back to the current issue, which is that gpr_native tracking is not perfect (as we may miss many patterns of spills/restores). I think we can use some heuristics like: if at a drreg spill, the reg is not native and and also not spilled to any drreg slot, then we assume that the gpr must be native right then. But I’m trying to think of a way to handle cases where we do have a spill slot at a spill but it is forgotten later.

example3:
```
<app2app spills app r1 to s1>
<app2app restores app r1 from s1>
<non-drreg r1 spill>
<non-drreg r1 restore> // now we wrongly think that r1 is !native
<insertion phase spills (app/tool?? Drreg won’t know) r1 to s2>
<insertion phase writes some tool val to r1>
<app2app spills app r2 to s1>
// at this point app r1 is only present in s2
```

Reply:
```
<spill reg r1 to s1>
<restore reg r1 from s1>
<restore reg r1 from s1>
<crash>
```

How does this sequence happen?  If the restores are before an app read, wouldn't they be accompanied by a tool spill and restore to a different slot?  So we can infer that the 1st refer is a lazy restore and r1 now has the app value?  And the 2nd restore can't happen w/o some intervening action?

Reply:
> If the restores are before an app read, wouldn't they either end a reservation or else be accompanied by a tool spill and restore to a different slot?

Yes, in this case, there is a tool val spill before the app read instr and after the app read instr a tool val restore. I omitted that here for brevity as the main question is still valid ("should we restore r1 from s1 after crash").

> So we can infer that the 1st refer is a lazy restore and r1 has the app value afterward?

Yes, the current logic does understand that r1 has the app value after a restore from s1. In this case, after the app read is complete, r1's tool value is restored, so we do need to restore r1's app value at crash.

> 2nd restore seems like it couldn't happen since r1 is not reserved?

r1 is still reserved. It was restored once only for the duration of the app read instr.


For reference: following is the instrumented sequence for `test_asm_fault_restore_gpr_restored_for_read`:

```
TAG  0x00007fb23802a183
 +0    L3 @0x00007faff81015c8  48 c7 c2 16 f1 00 00 mov    $0x000000000000f116 -> %rdx
 +7    L3 @0x00007faff80fdeb8  48 c7 c2 16 f1 00 00 mov    $0x000000000000f116 -> %rdx
 +14   m4 @0x00007faff80fffe8                       <label>
 +14   m4 @0x00007faff8176330  65 48 89 14 25 e0 00 mov    %rdx -> %gs:0x000000e0[8byte] // app spill
                               00 00
 +23   m4 @0x00007faff80e2768  48 c7 c2 cd ab 00 00 mov    $0x000000000000abcd -> %rdx // tool write
 +30   L3 @0x00007faff80e26e8  90                   nop
 +31   m4 @0x00007faff80dd8b8  65 48 89 14 25 e8 00 mov    %rdx -> %gs:0x000000e8[8byte] // tool spill
                               00 00
 +40   m4 @0x00007faff80ffee8  65 48 8b 14 25 e0 00 mov    %gs:0x000000e0[8byte] -> %rdx // app restore
                               00 00
 +49   L3 @0x00007faff80ff158  48 01 d3             add    %rdx %rbx -> %rbx
 +52   m4 @0x00007faff8107590  65 48 8b 14 25 e8 00 mov    %gs:0x000000e8[8byte] -> %rdx // tool restore
                               00 00
 +61   L3 @0x00007faff816bec8  48 c7 c1 00 00 00 00 mov    $0x0000000000000000 -> %rcx
 +68   L3 @0x00007faff8174990  48 8b 09             mov    (%rcx)[8byte] -> %rcx
 +71   m4 @0x00007faff8103360  65 48 8b 14 25 e0 00 mov    %gs:0x000000e0[8byte] -> %rdx // app restore
                               00 00
 +80   L4 @0x00007faff8108738  e9 7c 1e f3 3f       jmp    $0x00007fb23802a1a1
END 0x00007fb23802a183
```

Reply:
One of the complexities is that: we cannot know for sure which restore for a reg ends the reservation. Besides the final restore, there could be restores for app read (like the one above, which could arguably be discovered based on the extra tool spill and restore around it, but not sure if hard-coding that pattern is okay), and restores by the user to get the app value.

Reply:
> I omitted that here for brevity

I would claim that makes the question invalid: we would never see the sequence listed.  We would have the full info.

> be discovered based on the extra tool spill and restore around it, but not sure if hard-coding that pattern is okay

Since that is *our own pattern that we control*, then yes it seems ok IMHO.

> restores by the user to get the app value.

Ah this is pointing at something problematic.  So here is the example where we can't tell whether a restore ends the reservation: the user requesting a restore.  If this is the problem, maybe we could use a label here to help us identify it.

Reply:
> Since that is our own pattern that we control, then yes it seems ok IMHO.

Such patterns may be harder to detect if there are interleaving instrumentation between them added e.g. by the instru2instru phase.

Maybe we can use a combination of hardcoding patterns and adding labels to disambiguate restores and figure out where the gpr reservation ends, and then use the old logic of using the spill slot Id to differentiate between app and tool spills. I still need to think through some cases to convince myself. 

Aflags is a different story though. The most important difference being that it doesn't guarantee that re-spills will use the same slot, or any slot at all (if spilled to xax). So it's harder to differentiate between app aflags spill and tool aflags spill.

In my mind, my current approach of tracking the native value seemed better as other drreg logic can be completely decoupled from the restore state event.

Reply:
Update: I'm trying out an approach where, instead of walking the faulting fragment ilist from the start till the faulting instr, I walk it back from the end to the faulting instr. Finding the matching spill for a restore is turning out much simpler than finding the matching restore to a spill. This approach seems to work fine on the tests I have currently. Will send it over for review in a new PR when ready.

Reply:
Does the reverse walk make it easier to distinguish tool-value spills and user-requested restores?  I guess the final restore in the block always ends the reservation so that is what makes it easier?  Or is that always true: thinking of cases where a reg is live but then an app write makes it dead later, but there would still be a restore before the app read that made it live and that would be after all other restores.

Reply:
> Does the reverse walk make it easier to distinguish tool-value spills 

We want to see only app value restores and spills in this approach, and "jump" over tool value spills/restores: starting with the last restore[1] (which may be the end of block app val restore or one for an app read), we jump to its matching previous spill which is also guaranteed to be an app val spill. This goes on: any restore before this must also be an app value restore (or an app val write[2]), and so on. We backtrack until we find an app val spill-restore pair that crosses the faulting instr and use that slot to restore.

> user-requested restores

I see that `drreg_get_app_value` restores may be stateful if the reg is not in-use by drreg. So we don't want to ignore user restores every time. In any case, I think my current approach won't be affected by non-stateful restores: it doesn't care why a restore was added, just that it provided the app value for an app read (or end of block).

> I guess the final restore in the block always ends the reservation so that is what makes it easier?

Yes, but
[1]: I'm still debating whether I should start the reverse walk at the final restore in the block or the restore before the first app read after the faulting instr (if any). The latter feels better, but I'm unable to come up with a proof test case.

> thinking of cases where a reg is live but then an app write makes it dead later, but there would still be a restore before the app read that made it live and that would be after all other restores.

Right, there must be some app read, before the app write that makes it dead; or no app write at all in which case the reg is live beyond the block.


[2]: if the first app instr that accesses a reg after a fault is a write to that reg, then the reg is dead and doesn't need restoration. This shouldn't need special handling as we don't expect there to be any spill-restore pairs across an app write.

Reply:
> we don't expect there to be any spill-restore pairs across an app write.

Hmm.. this happens in the app2app scatter/gather expansion. We reserve aflags at the top and unreserve it at end, and there are `test` instrs (that write aflags) in the middle marked as app. On a separate note: it's weird the the corresponding `jz`s are not marked as app too; why the former but not the latter is marked as app?

I think this is an example that motivates starting the walk from the end of ilist, instead of first app read, so that we don't miss restore instrs. There may be unconventional uses in app2app phase that don't restore before app read or re-spill after app write.

Reply:
> There may be unconventional uses in app2app phase that don't restore before app read or re-spill after app write.

But then if some client makes such unconventional use, I think it's their responsibility to restore state on fault, right? And drx does that for the scatter/gather expansion.

Reply:
> On a separate note: it's weird the the corresponding jzs are not marked as app too; why the former but not the latter is marked as app?

A branch marked as app must end the block.  (DR then mangles it to have an exit stub and link it to other blocks.)